Warnings in `azure/identity.hpp` cause strict builds to fail

**Describe the bug**
I don't necessarily model this as a bug. This did affect my ability to build a project that had azure as a Third Party Dependency (https://github.com/apache/arrow/issues/39384), and so I wanted to let you know.

Essentially, `azure/identity.hpp` emitted a warning about visibility, that caused any strict build of Arrow to fail.

In the end Apache Arrow employed a workaround, where it ignores `-Wattributes` in `azure/identity.hpp`: https://github.com/apache/arrow/pull/39448/files

***Exception or Stack Trace***
```
[41/821] Building CXX object _deps/azure_sdk-build/sdk/identity/azure-identity/CMakeFiles/azure-identity.dir/src/client_certificate_credential.cpp.o                       
In file included from /home/anja/git/arrow/cpp/build/anja-python-debug/_deps/azure_sdk-src/sdk/identity/azure-identity/src/client_certificate_credential.cpp:4:            
/home/anja/git/arrow/cpp/build/anja-python-debug/_deps/azure_sdk-src/sdk/identity/azure-identity/inc/azure/identity/client_certificate_credential.hpp:68:9: warning: 'Azure
::Identity::ClientCertificateCredential' declared with greater visibility than the type of its field 'Azure::Identity::ClientCertificateCredential::m_pkey' [-Wattributes] 
   68 |   class ClientCertificateCredential final : public Core::Credentials::TokenCredential {                                                                            
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~                                                                                                                                
[45/821] Building CXX object _deps/azure_sdk-build/sdk/identity/azure-identity/CMakeFiles/azure-identity.dir/src/environment_credential.cpp.o                              
In file included from /home/anja/git/arrow/cpp/build/anja-python-debug/_deps/azure_sdk-src/sdk/identity/azure-identity/src/environment_credential.cpp:6:                   
/home/anja/git/arrow/cpp/build/anja-python-debug/_deps/azure_sdk-src/sdk/identity/azure-identity/inc/azure/identity/client_certificate_credential.hpp:68:9: warning: 'Azure
::Identity::ClientCertificateCredential' declared with greater visibility than the type of its field 'Azure::Identity::ClientCertificateCredential::m_pkey' [-Wattributes] 
   68 |   class ClientCertificateCredential final : public Core::Credentials::TokenCredential {     
   ```

**To Reproduce**
Unfortunately, the way I bumped into this was by attempting to build Arrow, and then it would install a bundled version of Azure SDK. I am hoping the stacktrace will help.

If it helps, this is my Arrow `CMakeUserPresets.json`:

```
{
    "version": 3,
    "cmakeMinimumRequired": {
      "major": 3,
      "minor": 21,
      "patch": 0
    },
    "configurePresets": [
        {
            "name": "user-base",
            "hidden": true,
            "binaryDir": "${sourceDir}/build/${presetName}",
            "cacheVariables": {
              "ARROW_INSTALL_NAME_RPATH": "OFF"
            }
        },
        {
            "name": "anja-python-debug",
            "inherits": ["ninja-debug-python", "features-filesystems", "user-base"],
            "cacheVariables": {
                "CMAKE_INSTALL_PREFIX": "/home/anja/miniconda3/envs/pyarrow-dev",
                "CMAKE_CXX_STANDARD": "17",
                 "Azure_SOURCE":"BUNDLED",
               "ARROW_BUILD_EXAMPLES": "ON",
                "ARROW_FILESYSTEM": "ON"
            }
        }
    ]
}
```

***Code Snippet***
See reproduction!

**Expected behavior**
No warnings.

**Setup (please complete the following information):**
 - OS: Ubuntu
 - IDE : vim
 - Version of the Library used: ARROW_AZURE_SDK_BUILD_VERSION=azure-core_1.10.3

**Information Checklist**
Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [X] Bug Description Added
- [X] Repro Steps Added
- [X] Setup information Added

--- a/sdk/identity/azure-identity/CHANGELOG.md
+++ b/sdk/identity/azure-identity/CHANGELOG.md
@@ -8,6 +8,8 @@
 
 ### Bugs Fixed
 
+- [[#5235]](https://github.com/Azure/azure-sdk-for-cpp/issues/5235) Warnings in `azure/identity.hpp` cause strict builds to fail.
+
 ### Other Changes
 
 ## 1.10.1 (2024-11-08)

--- a/sdk/identity/azure-identity/inc/azure/identity/client_certificate_credential.hpp
+++ b/sdk/identity/azure-identity/inc/azure/identity/client_certificate_credential.hpp
@@ -21,6 +21,14 @@
 #include <vector>
 
 namespace Azure { namespace Identity {
+
+#if defined(__GNUC__)
+// 'Azure::Identity::ClientCertificateCredential' declared with greater visibility than the type of
+// its field 'Azure::Identity::ClientCertificateCredential::m_pkey' [-Wattributes].
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wattributes"
+#endif
+
   namespace _detail {
     class TokenCredentialImpl;
 
@@ -174,5 +182,9 @@ namespace Azure { namespace Identity {
         Core::Credentials::TokenRequestContext const& tokenRequestContext,
         Core::Context const& context) const override;
   };
+
+#if defined(__GNUC__)
+#pragma GCC diagnostic pop
+#endif
 
 }} // namespace Azure::Identity

